src/app/InteractionModelEngine.cpp

```diff
@@ -485,6 +492,74 @@ void InteractionModelEngine::AddReadClient(ReadClient * apReadClient)
     mpActiveReadClientList = apReadClient;
 }
 
+bool InteractionModelEngine::CheckResourceQuotaForCurrentFabric(const ReadHandler * apReadHandler)
+{
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP && !CONFIG_IM_BUILD_FOR_UNIT_TEST
+    // If the resources are allocated on the heap, we should be able to handle as many Read / Subscribe requests as possible.
+    return true;
+#else
+    FabricIndex currentFabricIndex                     = apReadHandler->GetAccessingFabricIndex();
+    uint8_t fabricCount                                = mpFabricTable->FabricCount();
+    size_t attributePathsSubscribedByCurrentFabric     = 0;
+    size_t attributePathsUsedByCurrentFabric           = 0;
+    size_t eventPathsSubscribedByCurrentFabric         = 0;
+    size_t eventPathsUsedByCurrentFabric               = 0;
+    size_t dataVersionFiltersSubscribedByCurrentFabric = 0;
+    size_t dataVersionFiltersUsedByCurrentFabric       = 0;
+    size_t subscriptionsEstablishedByCurrentFabric     = 0;
+    size_t activeReadHandlersByCurrentFabric           = 0;
+
+    // It is safe to use & here since this function will be called on current stack.
+    mReadHandlers.ForEachActiveObject([&](ReadHandler * handler) {
+        if (handler->GetAccessingFabricIndex() == currentFabricIndex)
+        {
+            if (handler->IsType(ReadHandler::InteractionType::Subscribe))
+            {
+                attributePathsSubscribedByCurrentFabric += handler->GetAttributePathCount();
+                eventPathsSubscribedByCurrentFabric += handler->GetEventPathCount();
+                dataVersionFiltersSubscribedByCurrentFabric += handler->GetDataVersionFilterCount();
+
+                subscriptionsEstablishedByCurrentFabric++;
+            }
+            attributePathsUsedByCurrentFabric += handler->GetAttributePathCount();
+            eventPathsUsedByCurrentFabric += handler->GetEventPathCount();
+            dataVersionFiltersUsedByCurrentFabric += handler->GetDataVersionFilterCount();
+            activeReadHandlersByCurrentFabric++;
+        }
+        return Loop::Continue;
+    });
+
+    // Ensure the total number of subscriptions and ongoing read requests don't exceed the limit per fabric.
+    VerifyOrReturnError((subscriptionsEstablishedByCurrentFabric + 1) * fabricCount <= CHIP_IM_MAX_NUM_READ_HANDLER, false);
```

Comment:
Please document where the +1 is coming from.

Reply:
This function is refactored after this review comment